### PR TITLE
Add cos_zenith_angle_from_timestamp

### LINKED ISSUE #179: 🚀[FEA]: add cos zenith angle support for GPU
### Is this a new feature, an improvement, or a change to existing functionality?

Improvement

### How would you describe the priority of this feature request

Medium

### Please provide a clear description of problem you would like to solve.

[cos zenith angle](https://github.com/NVIDIA/modulus/blob/ccb7b6448f9f266763ce3a29bab69ec8c6a856f2/modulus/utils/sfno/zenith_angle.py#L51) cannot be run on the GPU since there is no date-time dtype on the gpu. A simple workaround is to modify the routine to work with UNIX timestamps in floating point or integer format. this is how google's graphcast handles time-dependent features. This will allow vectorizing the code across many times, using cuda graphs for models with cos zenith, compiling with jax and more.

I also noticed there are other cos zenith implementations in the code-base. Should these be consolidated in one namespace?

### Describe any alternatives you have considered

_No response_

### PR DESCRIPTION
**Pull request by nbren12**
# Modulus Pull Request

## Description

- Moves zenith angle code to the more vivible location `modulus.utils.zenith_angle`
- Add a function `modulus.utils.zenith_angle.cos_zenith_angle_from_timestamp` that works with floating point or other input rather than timestamps. This code could be used e.g. by jax to run on the GPU. It could be easily modified to add torch or cupy support, but I did not do that here. I needed this to allow including cos_zenith_angle in the jax jit compilation of the google graphcast in earth2mip
- add test of cos zenith angle from the original AI2 code base.

Closes #179 

## Checklist

- [ ] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/modulus/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/modulus/blob/main/CHANGELOG.md) is up to date with these changes.
- [x] An [issue](https://github.com/NVIDIA/modulus/issues) is linked to this pull request.

### CODE REVIEW

File: test/utils/test_zenith_angle.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    (
+        [datetime(2020, 3, 21, 12, 0, 0), 0.0, 0.0, 0.9994836252135212],
+        [datetime(2020, 3, 21, 18, 0, 0), -90.0, 0.0, 0.9994760971063111],
+        [datetime(2020, 3, 21, 18, 0, 0), 270.0, 0.0, 0.99947609879941],
+        [datetime(2020, 7, 6, 12, 0, 0), -90.0, 0.0, -0.019703903874316815],
+        [datetime(2020, 7, 6, 9, 0, 0), 40.0, 40.0, 0.9501802266240413],
+        [datetime(2020, 7, 6, 12, 0, 0), 0.0, 90.0, 0.3843918031907148],
+    ),
+)
+def test__sun_zenith_angle(time, lon, lat, expected):
```
Comment (mnabian) [inline test/utils/test_zenith_angle.py:49]: Is double underscore intentional?

File: test/utils/test_zenith_angle.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
+# ignore_header_test
+
+# climt/LICENSE
```
Comment (mnabian) [inline test/utils/test_zenith_angle.py:3]: Just checking if we should use the Apache license here? Or is this also third party code?

File: modulus/utils/zenith_angle.py [hunk 3]
```diff
@@ -74,10 +67,45 @@ def cos_zenith_angle(
...
+    --------
+    >>> model_time = datetime.datetime(2002, 1, 1, 12, 0, 0)
+    >>> angle = cos_zenith_angle(model_time.timestamp(), lat=360, lon=120)
+    >>> abs(angle - -0.447817277) < 1e-6
+    True
+    """
+    lon_rad = np.deg2rad(lon, dtype=dtype)
+    lat_rad = np.deg2rad(lat, dtype=dtype)
+    seconds_in_day = 86400
+    days_in_julian_year = 36525.0
```
Comment (akshaysubr) [inline modulus/utils/zenith_angle.py:101]: Why is this not `365.25` or the variable `days_in_julian_century`?

File: test/utils/test_zenith_angle.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+@pytest.mark.parametrize(
+    "time, lon, lat, expected",
+    (
+        [datetime(2020, 3, 21, 12, 0, 0), 0.0, 0.0, 0.9994836252135212],
+        [datetime(2020, 3, 21, 18, 0, 0), -90.0, 0.0, 0.9994760971063111],
+        [datetime(2020, 3, 21, 18, 0, 0), 270.0, 0.0, 0.99947609879941],
+        [datetime(2020, 7, 6, 12, 0, 0), -90.0, 0.0, -0.019703903874316815],
+        [datetime(2020, 7, 6, 9, 0, 0), 40.0, 40.0, 0.9501802266240413],
+        [datetime(2020, 7, 6, 12, 0, 0), 0.0, 90.0, 0.3843918031907148],
```
Comment (akshaysubr) [inline test/utils/test_zenith_angle.py:46]: Add a test for `T` being `np.ndarray`?

**Review by akshaysubr — CHANGES_REQUESTED** [review]
PR looks good. Had a couple of minor comments.

As an aside, this numba PR might be of interest as well: https://github.com/numba/numba/pull/3198

File: test/utils/test_zenith_angle.py (hunk 2 above)
Comment (nbren12) [reply test/utils/test_zenith_angle.py:3]: like the zenith angle implementation, this code is from https://github.com/ai2cm/fv3net/tree/master.

File: modulus/utils/zenith_angle.py (hunk 3 above)
Comment (nbren12) [reply modulus/utils/zenith_angle.py:101]: fixed.

File: test/utils/test_zenith_angle.py (hunk 1 above)
Comment (nbren12) [reply test/utils/test_zenith_angle.py:49]: changing to `test_zenith_angle`

File: test/utils/test_zenith_angle.py (hunk 4 above)
Comment (nbren12) [reply test/utils/test_zenith_angle.py:46]: added a test for to ensure the vectorization works:
```
    )


def test_zenith_angle_array():
    timestamp = np.array([0, 1, 2])[:, None, None]
    lat = np.array([0.0, 0.0])[None, :, None]
    lon = np.array([0.0])[None, None, :]
    out = cos_zenith_angle_from_timestamp(timestamp, lon, lat)
    assert out.shape == (3, 2, 1)

```

**Comment by nbren12** [comment]
Ready for re-review. I fixed the doctests.

**Review by mnabian — APPROVED** [review]

**Pull request merged**